Convert Gallina Extensions chapter to use prodns (#11705)
coq/coq: doc/sphinx/language/gallina-extensions.rst

.. _record_grammar:

.. cmd:: {| Record | Structure } @inductive_definition {* with @inductive_definition }

>    :name: Record; Structure

Reviewer: Since the recent changes, the first sentence of description doesn't make sense anymore. It would be good to fix it at the lowest possible cost (since this documentation is anyway going to be refactored soon).

Author: If the `type` must be a `sort`, should we update the mlg to reflect that?  Maybe applies in other points in the doc?

Author: Also, are both of these forms valid for a `Record`?  If not, should mention that.

![image](https://user-images.githubusercontent.com/1253341/76354430-ab4e2900-62cf-11ea-8f57-a77c1c3aa9e0.png)

Author: Check what I wrote--probably not entirely correct/complete yet.

Author: Can you have universes on the record `ident`?

Reviewer: - It doesn't have to be a `sort`. The grammar that you removed was artificially restrictive.

- For records, only the second construct is valid. The other one gives `The Record keyword is for types defined using the syntax { ... }.` Yes, it should be written down, and even the error message itself could be documented.

- If you mean the `ident` from the `ident { record_fields }` construct, then this is just the name of the record builder and its universes are likely to be exactly the same as the type itself, so there is probably no need to repeat the possibility of providing those universes.

Author: It seems to me that `inductive_definition` has been reused for several similar but distinct commands making the syntax a bit over-general.  Perhaps we should consider creating separate nonterminals in the doc, e.g. one for records and one for the case using constructors.  
Then the nonterminal for Record could omit the {? %| } {+| @constructor } entirely.

That would also simplify the first paragraph "Each inductive_definition defines a record...".

Reviewer: You can create inductive and co-inductive records with the `Inductive` or `CoInductive` keyword and the record syntax. That's why the two have been mixed. However, it should indeed be possible to separate the two syntax a bit more properly.

Reviewer: Cf. #11868. After digging into this, I am convinced that, at least at the documentation level, if not at the parsing level, we must do this split to better explain what is supported and what is not.

In particular, we should start by explaining `Variant` and `Record`, which both support a very limited syntax (they do support mutual definitions, but they make no sense, `Variant` only accept the constructors syntax and `Record` the record syntax).

Note that I'm talking about the refactored refman here, and all of this is out of scope for the present PR.

Reviewer: In the meantime, let's revert this description to `{| Record | Structure } {? > } @ident_decl {* @binder } {? : @type } {? := {? ident } { {*; record_field } } {? decl_notations } }` and let's move the definition of `record_field` and `field_body` here.

Author: Please check what I did for Record and Variant, close but slightly different from your recommendation.

   :name: Record; Structure

   .. insertprodn record_definition field_body

   .. prodn::
      record_definition ::= {? > } @ident_decl {* @binder } {? : @type } {? @ident } %{ {*; @record_field } %} {? @decl_notations }
      record_field ::= {* #[ {*, @attr } ] } @name {? @field_body } {? %| @num } {? @decl_notations }
      field_body ::= {* @binder } @of_type
      | {* @binder } @of_type := @term
      | {* @binder } := @term

   Each :n:`@record_definition` defines a record named by :n:`@ident_decl`.
   The constructor name is given by :n:`@ident`.
   If the constructor name is not specified, then the default name :n:`Build_@ident` is used,
   where :n:`@ident` is the record name.

   If :n:`@type` is
   omitted, the default type is :math:`\Type`. The identifiers inside the brackets are the field names.
   The type of each field :n:`@ident` is :n:`forall {* @binder }, @type`.
   Notice that the type of an identifier can depend on a previously-given identifier. Thus the
   order of the fields is important. :n:`@binder` parameters may be applied to the record as a whole
   or to individual fields.

>    Notations can be attached to fields using the :n:`decl_notations` annotation.

Reviewer: ```suggestion
   Notations can be attached to fields using the :n:`@decl_notations` annotation.
```

.. cmd:: {| Record | Structure } @inductive_definition {* with @inductive_definition }
   :name: Record; Structure

   Each :n:`@inductive_definition` defines a record named by the first :n:`@ident` within.
   The constructor name is the :n:`@ident` in the :n:`{? @ident } %{ {+; @record_field } %}` clause.
   If the constructor name is not specified, then the default name :n:`Build_@ident` is used,
   where :n:`@ident` is the record name.

   If :n:`@type` (which must be a :n:`sort`??) is
   omitted, the default sort is :math:`\Type`. The identifiers inside the brackets are the field names.
   The type of each field :n:`@ident`, its type is :n:`forall {* @binder }, @type`.
   Notice that the type of an identifier can depend on a previously-given identifier. Thus the
   order of the fields is important. :n:`@binder` parameters may be applied to the record as a whole

>    or to individual fields.

Reviewer: I feel like there could be fewer references to specific nonterminals that are not shown here and some generic explanations instead. But anyway, do not spend too much time making this perfect, since this is going to be moved closer to the other inductive definitions in the refactoring, and therefore, it should be easier to explain everything in order.

Author: Well, if you're trying to figure out the syntax for Record, you have to look at the definition of `inductive_definition` and some of the others.  It's important to be specific (there are multiple idents in the syntax).  At some points in the doc, I'm tempted to repeat some of the syntax just for easy reading (would require a small enhancement).  I think an astute reader would open the hyperlink in a new window so that they can see both the syntax and the text.

Alternatively and perhaps better, we could give a short example that the text could refer to.  If you give me an example covering the key points, I'll put it and adjust the text.  Other than that, not sure what I'd change.

What about ```(which must be a :n:`sort`??)```.

Author: > What about ```(which must be a :n:`sort`??)```.  

Ah, you covered this point above (it's not limited to a sort).

Reviewer: > But should I say `` the default sort is :math:`\Type`. `` or `` the default type is :math:`\Type`. ``?

Probably the latter.

>  At some points in the doc, I'm tempted to repeat some of the syntax just for easy reading

I think that the planned refactoring will help and you won't have to open several pages. It doesn't make sense that records are defined that late in the manual.

> The following experimental command is available when the ``FunInd`` library has been loaded via ``Require Import FunInd``:

Reviewer: We are well past this command being "experimental":

```suggestion
The following command is available when the ``FunInd`` library has been loaded via ``Require Import FunInd``:
```

.. cmd:: Function @fix_definition {* with @fix_definition }

   This command is a generalization of :cmd:`Fixpoint`. It is a wrapper
   for several ways of defining a function *and other useful related
   objects*, namely: an induction principle that reflects the recursive
   structure of the function (see :tacn:`function induction`) and its fixpoint equality.

>    This defines a function :n:`@ident` similar to :cmd:`Fixpoint`.

Reviewer: Since `ident` is not clearly defined above, I would simplify this to:

```suggestion
   This defines a function similar to :cmd:`Fixpoint`.
```

   As in :cmd:`Fixpoint`, the decreasing argument must
   be given (unless the function is not recursive), but it might not

>    necessarily be *structurally* decreasing. The point of the :n:`{ @fixannot }` annotation

Reviewer: ```suggestion
   necessarily be *structurally* decreasing. The point of the :n:`@fixannot` annotation
```

   is to name the decreasing argument *and* to describe which kind of
   decreasing criteria must be used to ensure termination of recursive
   calls.

:cmd:`Function` also supports the :n:`with` clause to create
mutually recursive definitions, however this feature is limited

> to structurally recursive functions.

Reviewer: ```suggestion
to structurally recursive functions, i.e. when n:`@fixannot` is a `struct`.
```

.. cmd:: Function @fix_definition {* with @fix_definition }

   This command is a generalization of :cmd:`Fixpoint`. It is a wrapper

>    for several ways of defining a function and other useful *related

Reviewer: I think the point was to insist on the *and* (it doesn't just define a function).

Author: I put it only on the "and".  What do you think?

BTW, the italic doesn't stand out very much.  Bold would definitely be noticeable.

![image](https://user-images.githubusercontent.com/1253341/76564718-5ee92180-6466-11ea-92ad-2ed9fa25db6a.png)

   clause).

   See :tacn:`function induction` and :cmd:`Functional Scheme` for how to use
   the induction principle to reason easily about the function.

   The form of the :n:`@fixannot` clause determines which definition mechanism :cmd:`Function` uses:

   *  If :n:`@fixannot` is not specified, :cmd:`Function`
      defines the nonrecursive function :token:`ident` as if it was declared with
      :cmd:`Definition`. In addition, the following are defined:

       + :token:`ident`\ ``_rect``, :token:`ident`\ ``_rec`` and :token:`ident`\ ``_ind``,
         which reflect the pattern matching structure of :token:`term` (see :cmd:`Inductive`);
       + The inductive :n:`R_@ident` corresponding to the graph of :token:`ident` (silently);
       + :token:`ident`\ ``_complete`` and :token:`ident`\ ``_correct`` which
         are inversion information linking the function and its graph.

   *  If :n:`{ struct ... }` is specified, :cmd:`Function`
      defines the structural recursive function :token:`ident` as if it was declared
      with :cmd:`Fixpoint`. In addition, the following are defined:

       + The same objects as above;

>        + The fixpoint equation of :token:`ident`: :n:`@ident`\_equation`.

Reviewer: ```suggestion
       + The fixpoint equation of :token:`ident`: :n:`@ident`\ ``_equation`.
```

       .. todo: these idents above are the not the one in { struct ident }, right?

>           AFAIK we don't describe the ident for struct anywhere

Reviewer: Indeed. And I guess this is simply because this behaves like `Fixpoint` so people can just refer to the explanation on `{ struct ... }` from the description of `Fixpoint`.

Author: Added a note here.

   *  If :n:`{ measure ... }` or :n:`{ wf ... }` are specified, :cmd:`Function`
      defines a recursive function by well-founded recursion. The module ``Recdef``
      of the standard library must be loaded for this feature.

>        + :n:`{measure @one_term {? @ident } {? @one_term } }`\: where :n:`@ident` is the decreasing argument

Reviewer: ```suggestion
       + :n:`{measure @one_term__1 {? @ident } {? @one_term__2 } }`\: where :n:`@ident` is the decreasing argument
```

(since this is referred below)

         and :n:`@one_term__1` is a function from the type of :n:`@ident` to :g:`nat` for which
         the decreasing argument decreases (for the :g:`lt` order on :g:`nat`)

>          for each recursive call of :n:`@one_term__1`. Parameters of the function are

Reviewer: ```suggestion
         for each recursive call of the function. The parameters of the function are
```

>          bound in :n:`@one_term`.

Reviewer: ```suggestion
         bound in :n:`@one_term__1`.
```

       + :n:`{wf @one_term @ident }`\: where :n:`@ident` is the decreasing argument and
         :n:`@one_term` is an ordering relation on the type of :n:`@ident` (i.e. of type
         `T`\ :math:`_{\sf ident}` → `T`\ :math:`_{\sf ident}` → ``Prop``) for which the decreasing argument

>          decreases for each recursive call of :n:`@one_term`. The order must be well-founded.

Reviewer: ```suggestion
         decreases for each recursive call of the function. The order must be well-founded.
```

>          Parameters of the function are bound in :n:`@one_term`.

Reviewer: ```suggestion
         The parameters of the function are bound in :n:`@one_term`.
```

>       If the annotation is ``measure`` or ``fw``, the user is left with some proof

Reviewer: ```suggestion
      If the annotation is ``measure`` or ``wf``, the user is left with some proof
```

      obligations that will be used to define the function. These proofs
      are: proofs that each recursive call is actually decreasing with
      respect to the given criteria, and (if the criteria is `wf`) a proof
      that the ordering relation is well-founded. Once proof obligations are
      discharged, the following objects are defined:

>         + The same objects as with the struct;

Reviewer: ```suggestion
        + The same objects as with the ``struct`` annotation;
```

>         + The lemma `ident`\ :math:`_{\sf tcc}` which collects all proof obligations in one

Reviewer: ```suggestion
        + The lemma :n:`@ident`\ ``_tcc`` which collects all proof obligations in one
```

          property;

>         + The lemmas `ident`\ :math:`_{\sf terminate}` and `ident`\ :math:`_{\sf F}` which is needed to be inlined

Reviewer: ```suggestion
        + The lemmas :n:`@ident`\ ``_terminate`` and :n:`@ident`\ ``_F`` which is needed to be inlined
```

Author: Got it.  I changed "which is needed to be inlined" to "which need to be inlined".  Is that good or is "which will be inlined" [by Coq] better?

Reviewer: I'd say the latter, but I'm not quite sure.

Author: Made it "will be".

>           during extraction of ident.

Reviewer: ```suggestion
          during extraction of :n:`@ident`.
```

      The way this recursive function is defined is the subject of several
      papers by Yves Bertot and Antonia Balaa on the one hand, and Gilles
      Barthe, Julien Forest, David Pichardie, and Vlad Rusu on the other
      hand. Remark: Proof obligations are presented as several subgoals

>       belonging to a Lemma `ident`\ :math:`_{\sf tcc}`.

Reviewer: This note seems completely useless and out of place to me.

```suggestion
      hand.
```

   See :tacn:`function induction` and :cmd:`Functional Scheme` for how to use
   the induction principle to reason easily about the function.

   The form of the :n:`@fixannot` clause determines which definition mechanism :cmd:`Function` uses.

>    (Note that references to :n:`ident` below refer to the :n:`ident` in the :n:`ident_decl`.):

Reviewer: ```suggestion
   (Note that references to :n:`ident` below refer to the :n:`ident` in the :n:`ident_decl` in the :n:`fix_definition`.):
```

> Depending on the ``{…}`` annotation, different definition mechanisms are

Reviewer: ```suggestion
Depending on the :n:`@fixannot` annotation, different definition mechanisms are
```

Above, it is written: "term must be built as a". This should be changed to something like "The body of the recursive definition must be built as a".

used by ``Function``. A more precise description is given below.

.. todo Théo, should the following info on cmdv:: Function s be moved adjacent

>    to the definition of @fixannot?  Maybe create a new subsection in the Gallina chapter?

Reviewer: No, because these are extensions of the core language.

The Gallina chapter should note that only the `{ struct }` annotation is part of the core language and that `Fixpoint` does not support `wf` and `measure` (indeed, when trying to use them, we get an error message "Well-founded induction requires Program Fixpoint or Function"). And that there use is documented at ...

Author: OK.  Can you give me a short description of when the user has to limit themself to non-extension features?  This can go in the introduction to the extensions chapter and briefly repeated where appropriate (such as for `struct`).

Reviewer: Never. It's just the approach we want to follow to explain things in the refactored manual: first explain what the kernel understands, then explain the rich language that Coq provides to the user and which is transformed to the core language of the kernel.

Author: Hmm, it seems like we should eventually explain that distinction and the reason for organizing the material that way (just for conceptual/pedagogic reasons).

Reviewer: Yes, absolutely, but that's not really the task of the current PR.

.. seealso:: :ref:`functional-scheme` and :tacn:`function induction`

> .. todo I think everything from here to "Section mechanism" should appear before the Note above

Reviewer: Sounds good!

Author: Moved it.  Our style seems to be to indent the description of commands and tactics, but it gets to be a long indentation, probably more than a page, which kinda defeats the purpose.  Perhaps worth reconsidering at some point whether such indentation is best.

   Most commands, like :cmd:`Hint`, :cmd:`Notation`, option management, … which
   appear inside a section are canceled when the section is closed.

.. cmd:: Let @ident @def_body

   This command binds the value :token:`term` to the name :token:`ident` in the

>    environment of the current section. The name :token:`ident` is accessible

Reviewer: This description is not completely exhaustive / accurate. It talks about `term` but `term` (the body) may not be provided, in which case the proof mode is opened. I suggest writing instead "The command works like `Definition` except that the declared constant is local to the section. When the section is closed..."

   appear inside a section are canceled when the section is closed.

.. cmd:: Let @ident @def_body
         Let Fixpoint @fix_definition {* with @fix_definition }
         Let CoFixpoint @cofix_definition {* with @cofix_definition }
   :name: Let; Let Fixpoint; Let CoFixpoint

>    These commands behave like :cmd:`Definition`, :cmd:`Fixpoint` and :cmd`CoFixpoint`, except that

Reviewer: ```suggestion
   These commands behave like :cmd:`Definition`, :cmd:`Fixpoint` and :cmd:`CoFixpoint`, except that
```

   the declared constant is local to the current section.
   When the section is closed, all persistent

>    definitions and theorems within it that depend on :token:`ident`

Reviewer: ```suggestion
   definitions and theorems within it that depend on the constant
```

>    will be prefixed with a :n:`term_let` with the same declaration.

Reviewer: ```suggestion
   will be prefixed with a :n:`@term_let` with the same declaration.
```

>    If :n:`@term` is omitted, :n:`@type` is required and Coq enters proof editing mode.

Reviewer: Can you add ```As for :cmd:`Definition`, :cmd:`Fixpoint` and :cmd:`CoFixpoint`,``` at the beginning of the paragraph?

   This can be used to define a term incrementally, in particular by relying on the :tacn:`refine` tactic.
   In this case, the proof should be terminated with :cmd:`Defined` in order to define a constant
   for which the computational behavior is relevant.  See :ref:`proof-editing-mode`.

   .. exn:: @ident already exists.
      :name: @ident already exists. (Let)

>       :undocumented:

Reviewer: IMHO there is no point in documenting this error message here since it is the same as in `Definition`, and since this description already heavily references the latter.

   Most commands, like :cmd:`Hint`, :cmd:`Notation`, option management, … which
   appear inside a section are canceled when the section is closed.

.. cmd:: Let @ident @def_body
         Let Fixpoint @fix_definition {* with @fix_definition }
         Let CoFixpoint @cofix_definition {* with @cofix_definition }
   :name: Let; Let Fixpoint; Let CoFixpoint

   These commands behave like :cmd:`Definition`, :cmd:`Fixpoint` and :cmd:`CoFixpoint`, except that
   the declared constant is local to the current section.
   When the section is closed, all persistent
   definitions and theorems within it that depend on the constant

>    will be prefixed with a :n:`@term_let` with the same declaration.

Author: How about "wrapped"?

>    will be prefixed with a :n:`@term_let` with the same declaration.

Reviewer: Sounds good.

      :name: @ident already exists. (Let)
      :undocumented:

>    .. cmd:: Let Fixpoint @fix_definition {* with @fix_definition }

Reviewer: I suggest moving this up to a shared block with `Let` as in:

```rst
.. cmd:: Let @ident @def_body
         Let Fixpoint @fix_definition {* with @fix_definition }
         Let CoFixpoint @cofix_definition {* with @cofix_definition }
```

Then the description becomes: "These commands behave like `Definition`, `Fixpoint` and `CoFixpoint` except that..."

Author: Does it make sense that `Let Fixpoint` and `Let CoFixpoint` allow a `univ_decl` on the `ident` but `Let` does not?

Reviewer: See #11534 for a beginning of discussion. I probably won't try to push this further on my own.

      | Module @qualid := @qualid
      module_expr_atom ::= @qualid
      | ( {+ @module_expr_atom } )
      of_module_type ::= : @module_type_inl
      | {* <: @module_type_inl }
      module_expr_inl ::= ! {+ @module_expr_atom }
      | {+ @module_expr_atom } {? @functor_app_annot }

   This command defines a module named :token:`ident`.  See the example :ref:`here<example_def_simple_module>`.

   The :n:`Import` and :n:`Export` flags specify whether the module should be automatically
   imported or exported.

   Specifying :n:`{* @module_binder }` starts an interactive functor with
   parameters given by the :n:`@module_binder`\s.

   .. todo: "interactive" means proof mode?

   :n:`@of_module_type` specifies the module type.  :n:`{+ <: @module_type_inl }`
   starts a module that satisfies each :n:`@module_type_inl`.

   :n:`:= {+<+ @module_expr_inl }` specifies the body for a module or functor definition.  If it's
   not specified, then the module is defined interactively.
   The ! prefix indicates that any
   `Inline` directive in the type of the functor arguments will be ignored
   (see the :cmd:`Module Type` command below).

>    .. todo: "interactively" above meaning in proof mode?  If so, maybe use that language

Reviewer: Not really "proof mode". Proof mode changes the parser to parse tactics and this is not the case here. However, it is similar in the sense that the definition of the module is between the `Module` and the `End` keywords instead of being after the `:=`.

   .. todo: What is an Inline directive?  sb command but still unclear.  Maybe referring to the
      "inline" in functor_app_annot?  or assumption_token inline assum_list?

.. cmd:: Module Type @ident {* @module_binder } {* <: @module_type_inl } {? := {+<+ @module_type_inl } }

   This command defines a module type named :n:`@ident`.

   Specifying :n:`{* @module_binder }` starts an interactive functor type with
   parameters given by the :n:`@module_binder`\s.

   :n:`:= ...` specifies a module or functor type definition.  If it's
   not specified, then the module is defined interactively.

.. note::

  #. Interactive modules and module types can be nested.
  #. Interactive modules and module types can't be defined inside of :ref:`sections<section-mechanism>`.
     Sections can be defined inside of interactive modules and module types.
  #. Hints and notations (:cmd:`Hint` and :cmd:`Notation` commands) can also appear inside interactive
     modules and module types. Note that with a module definition like:

  ::

     Module N : SIG := M.

  or::

>     Module N : SIG. … End N.

Author: Done, however, the `::` directive doesn't support `:n:` so I had to do something quite low-level.  In Latex, the second one will appear on a single line.  There's probably a `raw:: latex` that would do the trick but I'd rather not get into that now.

>     Module N : SIG. … End N.

Reviewer: I was suggesting to remove the `::` directive entirely and to inline this code in the paragraph.

  hints and the like valid for ``N`` are not those defined in ``M``
  (or the module body) but the ones defined in ``SIG``.

.. _terminating_module:

Terminating an interactive module or module type definition
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. todo this size is overkill for such a short section

.. todo: description below applies equally to modules and module types?

Interactive modules are terminated with the :cmd:`End` command, which
is also used to terminate :ref:`Sections<section-mechanism>`.
:n:`End @ident` closes the interactive module or module type :token:`ident`.
If the module type was given, the command verifies that the content of the module
matches the module type.  If the module is basic (not a
functor), its components (constants, inductive types, submodules etc.)
are now available through the dot notation.

.. exn:: No such label @ident.
    :undocumented:

.. exn:: Signature components for label @ident do not match.

>     :undocumented:

Reviewer: Another existing undocumented error is "The field @ident is missing in @qualid".

.. exn:: This is not the last opened module.
    :undocumented:

.. exn:: This is not the last opened module type.
   :undocumented:

> .. cmd:: Include @module_type_inl {* <+ @module_expr_inl }

Reviewer: ```suggestion
Commands to use inside the definition of an interactive module or module type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. cmd:: Include @module_type_inl {* <+ @module_expr_inl }
```

Without this sub-section title, it would seem like we are still within the "Terminating an interactive module..." sub-section. I also suggest that we swap the order of the two sub-sections.

Note that this is in this section that it should be explained that inside a module type definition `Parameter` et al. do not introduce a global axiom, but declare a constant that should be defined by any module of this type. Cf. the example of "module type" above.

Author: "Terminating an interactive module ..." is so short it doesn't deserve to be a section or in the TOC.  Also Include and Declare Module are accepted outside interactive definitions.

I tried to make this a subsubsection or a subsubsubsection but I couldn't figure out what characters to use to underline the title to make it so.  These are not right: https://documentation-style-guide-sphinx.readthedocs.io/en/latest/style-guide.html#headings.  If you know what they are, I'll add that info to doc/sphinx/README.md.

I moved the paragraph before the Note and put the title in bold.  Let me know what you think.

Author: I mentioned Parameter as item 4 in the Note.  Let me know if that covers it, I don't really know what you'd like to see where.  If this is not close, maybe you could give me some wording?

Reviewer: > "Terminating an interactive module ..." is so short it doesn't deserve to be a section or in the TOC.

Then, that's not the way to do it. The choice of characters for the underline doesn't change the level of the heading. You can choose anything and you just need to be consistent in your file.

>Also Include and Declare Module are accepted outside interactive definitions.

In truth, you are always in an interactive module definition, because files are modules themselves.

Author: Good.

> The choice of characters for the underline doesn't change the level of the heading.

That's not what the linked webpage says.  That's confusing.  How does one get subsubheaders?

Author: Here's an example from Gallina extensions of a header that doesn't appear in the TOC (and slightly smaller than a subsection header):

```
Implicit Argument Binders
+++++++++++++++++++++++++
```

![image](https://user-images.githubusercontent.com/1253341/77361501-9ed4c200-6d0c-11ea-8cca-ef3a37d50d65.png)

Reviewer: It's likely just because there was already many levels of headers above. In short, how to get a subsubheader: you need to put a section inside a section inside a section.

Author: Still a mystery to me, but we don't have to figure this out in this PR.

   Includes the content of module(s) in the current
   interactive module. Here :n:`@module_type_inl` can be a module expression or a module
   type expression. If it is a high-order module or module type
   expression then the system tries to instantiate :n:`@module_type_inl` with the current
   interactive module.

   Including multiple modules is a single :cmd:`Include` is equivalent to including each module
   in a separate :cmd:`Include` command.

.. cmd:: Include Type @module_type_inl {* <+ @module_type_inl }

>    .. deprecated:: Use :cmd:`Include` instead.

Reviewer: Actually `deprecated` wants to know since which version, so:

```suggestion
   .. deprecated:: 8.3
   
      Use :cmd:`Include` instead.
```

But since this has been deprecated and undocumented for so long, the best thing to do is probably to just remove it. See #11876.

Author: I would remove it when the command is removed.  Otherwise it will show up (to me, at least) as a undocumented command.  And it will be obvious to me when the command has been removed.

.. cmd:: Declare Module {? {| Import | Export } } @ident {* @module_binder } : @module_type_inl

   Declares a module :token:`ident` of type :token:`module_type_inl`.

   If :n:`@module_binder`\s are specified, declares a functor with parameters given by the list of
   :token:`module_binder`\s.

> .. _example_def_simple_module:

Reviewer: ```suggestion
Examples
~~~~~~~~

.. _example_def_simple_module:
```

Same thing as above. We need to show that we are out of the previous sub-section.

Author: I think the ambiguity of where subsections end is a general problem we should give some thought to.  We could eventually have a dividing line or center something like "* * *" for use where it helps.

How about if I move the example to the end of the module section?  There's already a hyperlink to the example in `Module`.

Reviewer: Yes to moving examples at the end. I think that's what I suggested somewhere.

.. example:: Defining a simple module

>     Let us define a simple module.

Reviewer: ```suggestion
```

This has become redundant with the title of the example.

BTW, this example sub-section would need some more `.. example::` blocks (I would say one for module type, one for functor, and one for the last example "module type with two sub-modules, sharing some of the fields").

Author: Reformatting the examples takes longer than you think.  Take a look.  Would be nice if coqtop would preserve indentation and blank lines, it could make the examples look better.

The module system provides a way of packaging related elements
together, as well as a means of massive abstraction.

In the syntax of module application, the ! prefix indicates that any
`Inline` directive in the type of the functor arguments will be ignored

> (see the :cmd:`Module Type` command below).

Author: Moved, but the wording needs work.

.. cmd:: Module {? {| Import | Export } } @ident {* @module_binder } {? @of_module_type } {? @is_module_expr }

   .. insertprodn module_binder module_expr_inl

   .. prodn::
      module_binder ::= ( {? {| Import | Export } } {+ @ident } : @module_type_inl )
      module_type_inl ::= ! @module_type
      | @module_type {? @functor_app_annot }
      module_type ::= @module_expr_atom
      | @module_type @module_expr_atom

>       | @module_type with @with_declaration

Author: Sorry, I had an incorrect edit here.  `module_type` should be:

```
      module_type ::= @qualid
      | ( @module_type )
      | @module_type @module_expr_atom
      | @module_type with @with_declaration
```

      with_declaration ::= Definition @qualid {? @univ_decl } := @term
      | Module @qualid := @qualid
      functor_app_annot ::= [ inline at level @num ]
      | [ no inline ]
      of_module_type ::= : @module_type_inl
      | {* <: @module_type_inl }
      module_expr_atom ::= @qualid
      | ( @module_expr )
      module_expr ::= @module_expr_atom
      | @module_expr @module_expr_atom

>       is_module_expr ::= := @module_expr_inl {* <+ @module_expr_inl }

Author: inlined

      module_expr_inl ::= ! @module_expr
      | @module_expr {? @functor_app_annot }

   This command starts an interactive module named :token:`ident`.

   The :n:`Import` and :n:`Export` flags specify the module should be automatically
   imported or exported.

   Specifying :n:`{* @module_binder }` starts an interactive functor with
   parameters given by the :n:`@module_binder`\s.

   :n:`@of_module_type` specifies the module type.  :n:`{+ <: @module_type_inl }`
   starts a module that satisfyies each :n:`@module_type_inl`.

   :n:`@is_module_expr` specifies the body for a module or functor definition.

.. todo: does the following title/organization make sense?  Should the info be given elsewhere??

>    Don't quite grok the difference between this and "an interactive module *type*"

Author: Please check what I did.

Reserved commands inside an interactive module
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. cmd:: Include @module_type_inl {* <+ @module_expr_inl }

   Includes the content of module(s) in the current
   interactive module. Here module can be a module expression or a module
   type expression. If module is a high-order module or module type
   expression then the system tries to instantiate module by the current
   interactive module.

.. cmd:: Include Type @module_type_inl {* <+ @module_type_inl }

> 

Reviewer: This was forgotten in the description of the `Include` command.

>    To be documented.

Reviewer: ```suggestion
   .. deprecated:: Use :cmd:`Include` instead.
```

.. cmdv:: End @ident

   This command closes the interactive module :token:`ident`. If the module type

>    was given the content of the module is matched against it and an error

Reviewer: ```suggestion
   was given, the content of the module is matched against it and an error
```

Interactive modules are terminated with the :cmd:`End` command, which
is also used to terminate :ref:`Sections<section-mechanism>`.
:n:`End @ident` closes the interactive module or module type :token:`ident`.
If the module type was given, the command verifies that the content of the module
matches the module type.  If the module is not a
functor, its components (constants, inductive types, submodules etc.)
are now available through the dot notation.

.. exn:: No such label @ident.
    :undocumented:

.. exn:: Signature components for label @ident do not match.
    :undocumented:

.. exn:: The field @ident is missing in @qualid.
   :undocumented:

.. |br| raw:: html

    <br>

.. note::

  #. Interactive modules and module types can be nested.
  #. Interactive modules and module types can't be defined inside of :ref:`sections<section-mechanism>`.
     Sections can be defined inside of interactive modules and module types.
  #. Hints and notations (:cmd:`Hint` and :cmd:`Notation` commands) can also appear inside interactive
     modules and module types. Note that with module definitions like:

     :n:`Module @ident__1 : @module_type := @ident__2.`

     or

> 

Reviewer: This is what I'm talking about.

Author: I think I got it.

You know I'm working primarily from the Conversation, not from the Files Changed.  It's not easy to get the context for a lot of your comments.  All I see is this (below), so it's not at all apparent that you're talking about the Inline clause of assumptions:

![image](https://user-images.githubusercontent.com/1253341/77366741-e3189000-6d15-11ea-9b28-66bdb71052c7.png)

   .. insertprodn module_binder module_expr_inl

   .. prodn::
      module_binder ::= ( {? {| Import | Export } } {+ @ident } : @module_type_inl )
      module_type_inl ::= ! @module_type
      | @module_type {? @functor_app_annot }
      functor_app_annot ::= [ inline at level @num ]
      | [ no inline ]
      module_type ::= @qualid
      | ( @module_type )
      | @module_type @module_expr_atom
      | @module_type with @with_declaration
      with_declaration ::= Definition @qualid {? @univ_decl } := @term
      | Module @qualid := @qualid
      module_expr_atom ::= @qualid
      | ( {+ @module_expr_atom } )
      of_module_type ::= : @module_type_inl
      | {* <: @module_type_inl }
      module_expr_inl ::= ! {+ @module_expr_atom }
      | {+ @module_expr_atom } {? @functor_app_annot }

   Defines a module named :token:`ident`.  See the examples :ref:`here<module_examples>`.

   The :n:`Import` and :n:`Export` flags specify whether the module should be automatically
   imported or exported.

   Specifying :n:`{* @module_binder }` starts a functor with
   parameters given by the :n:`@module_binder`\s.  (A *functor* is a function
   from modules to modules.)

>    .. todo: would like to find a better term than "interactive", not very descriptive

Reviewer: I agree but I don't have any suggestion.

Author: I thought about referring to "single-command module definition" and "multi-command module definition" which is descriptive but also a mouthful.

Reviewer: I wish the terminology was implying something about the use or not of the `:=` construct.

    :undocumented:

.. exn:: The field @ident is missing in @qualid.
   :undocumented:

.. |br| raw:: html

    <br>

.. note::

  #. Interactive modules and module types can be nested.
  #. Interactive modules and module types can't be defined inside of :ref:`sections<section-mechanism>`.
     Sections can be defined inside of interactive modules and module types.
  #. Hints and notations (:cmd:`Hint` and :cmd:`Notation` commands) can also appear inside interactive
     modules and module types. Note that with module definitions like:

     :n:`Module @ident__1 : @module_type := @ident__2.`

     or

     :n:`Module @ident__1 : @module_type.` |br|
     :n:`Include @ident__2.` |br|
     :n:`End @ident__1.`

     hints and the like valid for :n:`@ident__1` are the ones defined in :n:`@module_type`
     rather then those defined in :n:`@ident__2` (or the module body).
  #. Within an interactive module type definition, the :cmd:`Parameter` command declares a
     constant instead of definining a new axiom (which it does when not in a module type definition).
  #. Assumptions such as :cmd:`Axiom` that include the :n:`Inline` clause will be automatically

>      expanded when the functor is applied, except when the function application is prefixed by ``!``.

Reviewer: Looks good except that it can be confusing that the two successive notes do not refer to the same assumption command (`Parameter` / `Axiom`) even though these are equivalent. Not a big deal either.

Author: It's particularly odd to say "Parameter ... instead of defining a new **axiom**" if Parameters and Axioms are the same thing.  The equivalence deserves more explanation under the Parameter command.

The wording for Parameter should include Conjecture and Axiom.  And not clear if "Assumptions such as ..." include Variable and Hypothesis.

Please elaborate or I'll leave it as-is.

Reviewer: Yes, these are the same thing, but yes, we tend to use a different terminology because they are not meant to be used in the same cases. But I think it is also fine to leave as is.

   Declares a module :token:`ident` of type :token:`module_type_inl`.

   If :n:`@module_binder`\s are specified, declares a functor with parameters given by the list of
   :token:`module_binder`\s.

.. cmd:: Import {+ @qualid }

   If :token:`qualid` denotes a valid basic module (i.e. its module type is a
   signature), makes its components available by their short names.

   .. example::

      .. coqtop:: reset in

         Module Mod.
         Definition T:=nat.
         Check T.
         End Mod.
         Check Mod.T.

      .. coqtop:: all

         Fail Check T.
         Import Mod.
         Check T.

   Some features defined in modules are activated only when a module is
   imported. This is for instance the case of notations (see :ref:`Notations`).

   Declarations made with the :attr:`local` attribute are never imported by the :cmd:`Import`
   command. Such declarations are only accessible through their fully
   qualified name.

   .. example::

      .. coqtop:: in

         Module A.
         Module B.
         Local Definition T := nat.
         End B.
         End A.
         Import A.

>    .. coqtop:: all fail

Reviewer: This `coqtop` block is not indented enough and appears out of the example block even though it is fully relevant to the example.

Author: Indented but still doesn't look right.

      :undocumented:

.. cmd:: Print Module @qualid

   Prints the module type and (optionally) the body of the module :n:`@qualid`.

.. cmd:: Print Module Type @qualid

   Prints the module type corresponding to :n:`@qualid`.

.. flag:: Short Module Printing

   This flag (off by default) disables the printing of the types of fields,
   leaving only their names, for the commands :cmd:`Print Module` and
   :cmd:`Print Module Type`.

.. _module_examples:

Examples
~~~~~~~~

.. example:: Defining a simple module interactively

    .. coqtop:: in

       Module M.
       Definition T := nat.
       Definition x := 0.

    .. coqtop:: all

       Definition y : bool.
       exact true.

    .. coqtop:: in

       Defined.
       End M.

Inside a module one can define constants, prove theorems and do anything
else that can be done in the toplevel. Components of a closed
module can be accessed using the dot notation:

.. coqtop:: all

   Print M.x.

.. _example_def_simple_module_type:

.. example:: Defining a simple module type interactively

   .. coqtop:: in

      Module Type SIG.
      Parameter T : Set.
      Parameter x : T.
      End SIG.

.. _example_filter_module:

> .. example Creating a new module that omits some items from an existing module

Reviewer: ```suggestion
.. example:: Creating a new module that omits some items from an existing module
```

Author: Yup.  Maybe worth creating a script that would grep for problems like this, for example reporting instances of `\.\. example[^:]`.

   Since the type, :n:`SIG`, of the new module :n:`N` doesn't define y or give the body of :n`x`,
   those are not included in :n:`N`.

   .. coqtop:: all

      Module N : SIG with Definition T := nat := M.
      Print N.T.
      Print N.x.
      Fail Print N.y.

.. reset to remove N (undo in last coqtop block doesn't seem to do that), invisibly redefine M, SIG
.. coqtop:: none reset

   Module M.
   Definition T := nat.
   Definition x := 0.
   Definition y : bool.
   exact true.
   Defined.
   End M.

   Module Type SIG.
   Parameter T : Set.
   Parameter x : T.
   End SIG.

The following definition of :g:`N` using the module type expression :g:`SIG` with
:g:`Definition T := nat` is equivalent to the following one:

.. todo: what is other definition referred to above?

>    "Module N' : SIG with Definition T := nat. End N`." is not it.

Reviewer: I don't know. These examples will need to be reworked, but maybe not now. Please do leave your `.. todo` note in the code.

      module_expr_inl ::= ! {+ @module_expr_atom }
      | {+ @module_expr_atom } {? @functor_app_annot }

   Defines a module named :token:`ident`.  See the example :ref:`here<example_def_simple_module>`.

   The :n:`Import` and :n:`Export` flags specify whether the module should be automatically
   imported or exported.

   Specifying :n:`{* @module_binder }` starts an interactive functor with
   parameters given by the :n:`@module_binder`\s.  (A *functor* is a module

>    with parameters.)

Author: Got the "interactive".

We need to describe functors somewhere.  Eventually (or now?).  Not sure they're even well-covered in the OCaml doc.  Can you have a useful functor without parameters?  Also, your use of "function" is in the mathematical sense rather the common, specific term "function" in Coq or OCaml, right?  Needs more words to be clear.  Maybe add some now?

>    with parameters.)

Reviewer: Right.

In any case, I believe that major writing improvements to the documentation of modules should be delayed to the moment that we merge this section with the current Modules chapter (which contains the typing rules) into a shiny new section of the Core language chapter.

>    with parameters.)

Author: I will leave it as "(A *functor* is a function from modules to modules.)"

   Interactively defining the :n:`@module_expr_inl`\s in a series of
   :cmd:`Include` commands is equivalent to giving them all in a single
   non-interactive :cmd:`Module` command.

   The ! prefix indicates that any
   `Inline` directive in the type of the functor arguments will be ignored.

   .. todo: What is an Inline directive?  sb command but still unclear.  Maybe referring to the

>       "inline" in functor_app_annot?  or assumption_token Inline assum_list?

Reviewer: I think the latter. Indeed, this documentation of the `!` prefix clearly mirrors the one of the `Inline` clause.

Author: This applies only to assumptions and not to these?

```
    Extraction Inline {+ @qualid }
    Print Extraction Inline
    Register Inline @qualid
    Reset Extraction Inline
```

Reviewer: Interesting. I guess maybe `Register` as well, probably not the others. This would be better if it was an attribute. Anyway, to leave for further consideration.

    .. exn:: This is not the last opened module.
       :undocumented:

> 

Reviewer: This was forgotten in the description of `Module`.

Author: While there is a direct relationship, how can I clearly and simply state how they're equivalent (e.g. at a syntactic level).  Interactively the `module_expression`s would appear in different commands--what are those commands?

Not sure this is really that important a point to make.

Reviewer: >  Interactively the `module_expression`s would appear in different commands--what are those commands?

`Include`

> Not sure this is really that important a point to make.

I'm just ensuring we are not losing documentation while making these changes.

Reserved commands inside an interactive module type:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: The special meaning of `Parameter` *et al.* in interactive module types should be explained in the section on "Commands inside interactive modules and module types" even if the command is not shown here because it was already defined in the Gallina chapter. In particular, the meaning of the `Inline` annotation which was not explained in the previous chapter should still be explained here. In the previous chapter, there should be a mention that these commands take a (slightly) different meaning when inside interactive module types, and in particular that the `Inline` annotation is relevant only in this case.

Author: Sounds terrific, but I have little idea what should be written on this here or elsewhere.  You mean to add a new section "Commands inside ..."?  I suppose the paragraph below the `Parameter` description containing "Outside of a section, these are equivalent to :n:`Local Parameter`." should say "section or module"?

By "`Inline` annotation", you mean the `Inline` clause in Parameter et. al.?  IMHO "clause" is better because it more clearly implies syntax, whereas "annotation" and some similar words sound like they may have specific definitions within Coq.  No idea what `Inline` does here.

If you give me the facts in some detail (outline is fine), I will come up with the language.  Which ever way you prefer.

Reviewer: > You mean to add a new section "Commands inside ..."?

I meant that the section "Reserved commands inside..." could simply be renamed "Commands inside..." to include commands that are not reserved but change meaning inside interactive modules / module types.

> I suppose the paragraph below the `Parameter` description containing "Outside of a section, these are equivalent to :n:`Local Parameter`." should say "section or module"?

No, I don't think that needs changing.

>  IMHO "clause" is better because it more clearly implies syntax

Yes.

> No idea what `Inline` does here.

Well, it is documented here, even though I cannot yet decipher what this means myself. For sure, we'll need to improve this once we start really working on module documentation (in particular with the merging of the module chapter and this section).

Author: Then if you're good with this part as it is, please resolve.  Or elaborate.

Reviewer: The only thing remaining is the `Inline` documentation which is also tracked in other comments below.

Author: Think I got it.

.. cmdv:: End @ident

   This command closes the interactive module type :token:`ident`.

   .. exn:: This is not the last opened module type.
      :undocumented:

.. cmd:: Module Type @ident {* @module_binder } {+ <: @module_type_inl } {? @is_module_type }

   .. insertprodn is_module_type is_module_type

   .. prodn::

>       is_module_type ::= := @module_type_inl {* <+ @module_type_inl }

Reviewer: I suggest inlining this or renaming to `module_type_body` (but I think I prefer the former).

>    This command is used to start an interactive module type :n:`@ident`.

Reviewer: (unless the body is provided)

Author: Put this with different words a bit lower in the description.

   Specifying :n:`{* @module_binder }` starts an interactive functor type with
   parameters given by the :n:`@module_binder`\s.

> 

Reviewer: This was forgotten in the description of `Module Type`.

Author: Same idea; see previous comment https://github.com/coq/coq/pull/11705#discussion_r396039668

Author: IIUC `Include` and `Include Type` in an interactive definition are equivalent to giving the same definitions in "Module" and "Module Type" respectively.  How would `Declare Module` in an interactive definition fit into a single-line module definition?  Is it true that everything you can do interactively has an equivalent single-command definition?  From your earlier example:
```
Module Type contains_empty (empty_arg : empty_type).
Declare Module empty : empty_type.
End contains_empty.
```

I can leave a `todo` here if this is getting in too deep here for this PR.

Reviewer: > IIUC `Include` and `Include Type` in an interactive definition are equivalent to giving the same definitions in "Module" and "Module Type" respectively

That may have been the idea originally but now they are strictly equivalent commands.

>  Is it true that everything you can do interactively has an equivalent single-command definition?

No, I don't think so. Not in the case of modules.

.. exn:: Signature components for label @ident do not match.
    :undocumented:

.. exn:: The field @ident is missing in @qualid.
   :undocumented:

.. |br| raw:: html

    <br>

.. note::

  #. Interactive modules and module types can be nested.
  #. Interactive modules and module types can't be defined inside of :ref:`sections<section-mechanism>`.
     Sections can be defined inside of interactive modules and module types.
  #. Hints and notations (:cmd:`Hint` and :cmd:`Notation` commands) can also appear inside interactive
     modules and module types. Note that with module definitions like:

     :n:`Module @ident__1 : @module_type := @ident__2.`

     or

     :n:`Module @ident__1 : @module_type.` |br|
     :n:`Include @ident__2.` |br|
     :n:`End @ident__1.`

     hints and the like valid for :n:`@ident__1` are the ones defined in :n:`@module_type`
     rather then those defined in :n:`@ident__2` (or the module body).

  #. Within an interactive module type definition, the :cmd:`Parameter` command declares a

>      constant instead of definining a new axiom (which it does when not in a module type definition).

Reviewer: Looks good but the documentation of the `Inline` clause should also be put here.

Author: Unclear what you want me to add (or move?) here.

Reviewer: Cf. the comment just below "This is what I'm talking about".

> .. cmd:: Include @module_type_inl {* <+ @module_expr_inl }

Reviewer: Since in the syntax of `Module` and `Module Type` use `{+<+ @module_expr_inl}`, I think it would make sense that the `Include` command uses the same, i.e.
```suggestion
.. cmd:: Include {+<+ @module_expr_inl }
```

Author: Really?? module_type_inl and module_type_expr are not the same thing.  I will do it for `Include Type`, though.

Reviewer: I hadn't realized these were not the same non-terminals. I wonder why it is so, given the description.

.. _example_filter_module:

.. example:: Creating a new module that omits some items from an existing module

   Since :n:`SIG`, the type of the new module :n:`N`, doesn't define :n:`y` or

>    give the body of :n:`x`, which are not included in :n:`N`.

Reviewer: There's an inconsistency between here and the use of `:g:` below. At some point, we should clarify what kind of verbatim quoting we want to keep and in which case we should use which.

   .. example::

      .. coqtop:: reset in

         Module Mod.
         Definition T:=nat.
         Check T.
         End Mod.
         Check Mod.T.

>       .. coqtop all

Reviewer: ```suggestion
      .. coqtop:: all
```

Notice that ``M`` is a correct body for the component ``M2`` since its ``T``
component is equal ``nat`` and hence ``M1.T`` as specified.

> .. _import_qualid:

Reviewer: Is this label actually meaningful?

Author: removed

> .. cmd:: Import {+ @qualid }

Reviewer: `Import` and `Export` could be moved to the previous section, while the printing commands could  be moved to the terminating a module section (I guess).

Author: Let's figure out the terminating a module section.  `Export` is accepted outside a module.  Is that actually useful or a curio?

Reviewer: As I said somewhere above, as soon as you open a file, you are actually interactively defining a module.

Author: I moved the examples to the very end of the chapter, looks good to me.

   .. example::

      .. coqtop:: in

         Module A.
         Module B.
         Local Definition T := nat.
         End B.
         End A.
         Import A.

   .. coqtop:: all

>          Fail Check B.T.

Reviewer: This could be also the following to avoid the need for `Fail`:

```suggestion
   .. coqtop:: all fail

         Check B.T.
```

         Fail Check B.T.

   .. cmd:: Export {+ @qualid }
      :name: Export

>       When the module containing the command ``Export`` qualid

Reviewer: ```suggestion
      Same thing as :cmd:`Import` except that when the module containing this command
```

line below is:

```rst
      is imported, qualid is imported as well.
```

and should become:

```rst
      is imported, :n:`{+ @qualid }` are imported as well.
```

The command line option ``-R`` is a variant of ``-Q`` which has the strictly
same behavior regarding loadpaths, but which also makes the
corresponding ``.vo`` files available through their short names in a way

> not unlike the ``Import`` command (see :cmd`Import`). For instance, ``-R path Lib``

Reviewer: ```suggestion
not unlike the :cmd:`Import` command. For instance, ``-R path Lib``
```

      smart_qualid ::= @qualid
      | @by_notation
      argument_spec_block ::= @argument_spec
      | /
      | &
      | ( {+ @argument_spec } ) {? % @ident }
      | [ {+ @argument_spec } ] {? % @ident }
      | %{ {+ @argument_spec } %} {? % @ident }
      argument_spec ::= {? ! } @name {? % @ident }
      more_implicits_block ::= @name
      | [ {+ @name } ]
      | %{ {+ @name } %}
      arguments_modifier ::= simpl nomatch
      | simpl never
      | default implicits
      | clear bidirectionality hint
      | clear implicits
      | clear scopes
      | clear scopes and implicits
      | clear implicits and scopes
      | rename
      | assert
      | extra scopes

   This command sets implicit arguments *a posteriori*,
   where the list of :n:`@name`\s is a prefix of the list of
   arguments of :n:`@qualid` (in the :n:`@smart_qualid`).  Arguments in square
   brackets are declared as implicit and arguments in curly brackets are declared as
   maximally inserted.

   After the command is issued, implicit arguments can and must be
   omitted in any expression that applies :token:`qualid`.

   This command supports the :attr:`local` and :attr:`global` attributes.

>    Default behavior is to limit the affect to the current section but also to

Reviewer: ```suggestion
   Default behavior is to limit the effect to the current section but also to
```

.. cmd:: Arguments @smart_qualid {* @argument_spec_block } {* , {* @more_implicits_block } } {? : {+, @arguments_modifier } }
   :name: Arguments

   .. insertprodn smart_qualid arguments_modifier

   .. prodn::
      smart_qualid ::= @qualid

>       | @by_notation

Author: No worries, an easy change.

   :name: Arguments

   .. insertprodn argument_spec_block arguments_modifier

   .. prodn::
      argument_spec_block ::= @argument_spec
      | /
      | &
      | ( {+ @argument_spec } ) {? % @ident }
      | [ {+ @argument_spec } ] {? % @ident }
      | %{ {+ @argument_spec } %} {? % @ident }
      argument_spec ::= {? ! } @name {? % @ident }
      more_implicits_block ::= @name
      | [ {+ @name } ]
      | %{ {+ @name } %}
      arguments_modifier ::= simpl nomatch
      | simpl never
      | default implicits
      | clear bidirectionality hint
      | clear implicits
      | clear scopes
      | clear scopes and implicits
      | clear implicits and scopes
      | rename
      | assert
      | extra scopes

   This command sets implicit arguments *a posteriori*,
   where the list of :n:`@name`\s is a prefix of the list of
   arguments of :n:`@qualid` (in the :n:`@smart_qualid`).  Arguments in square
   brackets are declared as implicit and arguments in curly brackets are declared as
   maximally inserted.

   After the command is issued, implicit arguments can and must be
   omitted in any expression that applies :token:`qualid`.

>    This command supports the :attr:`local` and :attr:`global` attributes.

Author: FWIW AFAIK it was not explained in this file, but only as cmdvs in syntax-extensions.rst and under :attr:`local`.  I added the info here nonetheless, but still not exhaustive or all that clear.  Please take a look.

>    This command supports the :attr:`local` and :attr:`global` attributes.

Reviewer: No, it was explained here. Otherwise, I would not have left this comment:

```rst
 .. cmdv:: Global Arguments @qualid {* {| [ @name ] | { @name } | @name } }

    This command is used to recompute the implicit arguments of
    :token:`qualid` after ending of the current section if any, enforcing the
    implicit arguments known from inside the section to be the ones
    declared by the command.

.. cmdv:: Local Arguments @qualid {* {| [ @name ] | { @name } | @name } }

    When in a module, tell not to activate the
    implicit arguments of :token:`qualid` declared by this command to contexts that
    require the module.
```

That being said, this explanation was quite undecipherable and your addition is better IMHO.

   A command containing :n:`@argument_spec_block & @argument_spec_block`
   provides :ref:`bidirectionality_hints`.

   Use the :n:`@more_implicits_block` to specify multiple implicit arguments declarations
   for names of constants, inductive types, constructors and lemmas that can only be
   applied to a fixed number of arguments (excluding, for instance,
   constants whose type is polymorphic).
   The longest applicable list of implicit arguments will be used to select which
   implicit arguments are inserted.
   For printing, the omitted arguments are the ones of the longest list of implicit
   arguments of the sequence.  See the example :ref:`here<example_more_implicits>`.

   The :n:`@arguments_modifier` values have various effects:

   * :n:`clear implicits` - clears implicit arguments
   * :n:`default implicits` - automatically determine the implicit arguments of the object.
     See :ref:`auto_decl_implicit_args`.
   * :n:`rename` - rename implicit arguments for the object
   * :n:`assert` - assert that the object has the expected number of arguments with the
     expected names.  See the example here: :ref:`renaming_implicit_arguments`.

.. _example_more_implicits:

> .. example:: Multiple implicit arguments with :n:`@more_implicits_block`

Author: You'd put that title just above `Arguments map [A B] f l, [A] B f l, A B f l.`?

Would you suggest titles for rest of the example?  I don't know what titles to give them.

> .. example:: Multiple implicit arguments with :n:`@more_implicits_block`

Reviewer: Yes, and I'm not sure. In doubt, don't give titles to the others.

  :undocumented:

.. cmd:: Canonical {? Structure } @by_notation
   :name: _

   The first form of this command declares an existing :n:`@qualid` as a

>    canonical instance of a structure (a record).

Reviewer: This description makes no sense (in the current state) because you applied it without applying the changes to the command syntax that I suggested.

Author: Already fixed.

Also note the comments in the Conversation section are time-ordered.  I addressed this point a couple hours ago in two earlier comments.  Doesn't make it any easier.  And so many details.

value. The complete documentation of canonical structures can be found
in :ref:`canonicalstructures`; here only a simple example is given.

.. cmd:: Canonical {? Structure } @qualid {? {? @univ_decl } @def_body }
  :name: Canonical Structure
  :undocumented:

.. cmd:: Canonical {? Structure } @by_notation

>    :name: Canonical Structure X

Author: You want me to mention that in the text?  BTW, the message appears in the code but doesn't appear anywhere in the doc.  I could add it here or maybe wherever qualid is defined.

>    :name: Canonical Structure X

Reviewer: >You want me to mention that in the text?

No. That's really an implementation detail working around a limitation of what can be done with a parser.

>    :name: Canonical Structure X

Author: What about adding "This expression should be a simple identifier." somewhere?

>    :name: Canonical Structure X

Author: Should this be indexed under "Canonical" or "Canonical Structure", or both?

>    :name: Canonical Structure X

Reviewer: Just use `ident_decl` like in my suggestion above and do not write anything about it. I just explained the reason I did this transformation, I was not suggesting any change beyond the content of my "suggestion".

I think just "Canonical Structure" is fine.

>    This command declares :token:`qualid` as a canonical instance of a

Author: qualid, not smart_qualid

>    This command declares :token:`qualid` as a canonical instance of a

Reviewer: Why not `smart_qualid`? Cf. my suggested change just above.

>    This command declares :token:`qualid` as a canonical instance of a

Author: OK, after staring at it for 20 minutes, I think I understand your change above.  Not a very obvious transformation.  One almost wants a proof of correctness.  Will do both changes.

>    This command declares :token:`qualid` as a canonical instance of a

Reviewer: OK great. Indeed, I could have explained it more.

>    This command declares :token:`qualid` as a canonical instance of a

Reviewer: The sad thing is that my change is certainly what the programmer wanted, and they had to work around parsing issues and implement it in a twisted way, and then we had to retro-engineer it.

value. The complete documentation of canonical structures can be found
in :ref:`canonicalstructures`; here only a simple example is given.

.. cmd:: Canonical {? Structure } @smart_qualid
         Canonical {? Structure } @ident_decl @def_body

>    :name: Canonical Structure; _

Author: You're referring to the permalink, right?  A bug indeed.

![image](https://user-images.githubusercontent.com/1253341/77461482-dc465780-6dbf-11ea-8121-d8945d7ab4a9.png)

>    :name: Canonical Structure; _

Reviewer: Yes.

      only the structure declared first as canonical is considered.

   .. attr:: canonical(false)

>       :name: canonical(false)

Author: If it was a command or tactic, the :name: would be required.  I may make attr behave this way for consistency later on (indeed, I have a branch with that behavior already)

>       :name: canonical(false)

Reviewer: Because of the parentheses?

>       :name: canonical(false)

Author: Correct.  `cmd:: foo ( )` is indexed by default under `foo`.  The default name code also replaces notation stuff with `...`.

>       :name: canonical(false)

Reviewer: OK thanks for the explanation.

       Lemma cons_inj_bool : forall (m n:bool) l, n :: l = m :: l -> n = m.
       Abort.

.. cmd:: Implicit Type @reserv_list

  This is useful for declaring the implicit type of a single variable.

>   .. todo then why does it accept multiple values?

Author: Combined.  Let me know if you want to index to show just "Implicit" or "Implicit Type"+"Implicit Types"

>   .. todo then why does it accept multiple values?

Reviewer: The latter.

   Check (forall `{Commutative A, Hnat : !Commutative nat}, True).

.. cmd:: Generalizable {| All Variables | No Variables | {| Variable | Variables } {+ @ident } }

   Controls the set of generalizable identifiers to avoid unexpected generalizations

>    from typos in identifiers.   The effect of multiple :cmd:`Generalizable` commands

Reviewer: This sentence is mostly intended to describe the "Variable(s)" variant *as opposed to* the "All Variables" variant (one of them being needed to use the feature).

>    is cumulative; each command adds additional generalizable :n:`@ident`\s.

Reviewer: This second sentence is mostly applicable to the "Variable(s)" variants. The "All Variables" and "No Variables" variant completely wipe out the effect of the previous commands.

Author: OK.  Reordered the options to make it read better.

generalizations when mistyping identifiers. There are several commands
that specify which variables should be generalizable.

.. cmd:: Generalizable {| All Variables | No Variables | {| Variable | Variables } {+ @ident } }

>    All variables are candidates for

Reviewer: This is the description for the `All Variables` variant. This should be made explicit.

Author: Check what I wrote.  I have no idea what a "generalization delimiter" is.

Reviewer: This likely refers to the ``` `( @ident : @type ) ``` / ``` `{ @ident : @type } ``` / ``` `[ @ident : @type ] ``` from `binder` (I simplified the syntax on purpose).

Reviewer: BTW, I think we should move this part of the `binder` syntax into a new nonterminal `generalizing_binder` to be defined around here. To be clear I am talking about:

![2020-03-23-114051_468x245_scrot](https://user-images.githubusercontent.com/1108325/77308388-29acc100-6cfb-11ea-9575-17a2757d84d9.png)

And we should also move the two rules just above:

![2020-03-23-114151_472x64_scrot](https://user-images.githubusercontent.com/1108325/77308491-506af780-6cfb-11ea-959d-c110ad239fa5.png)

to a new nonterminal `implicit_binders` to be defined in the section on "Implicit Argument Binders".

   generalization if they appear free in the context under a
   generalization delimiter. This may result in confusing errors in case
   of typos. In such cases, the context will probably contain some
   unexpected generalized variable.

>    The :n:`No Variables` option disables implicit generalization entirely.  This is the

Reviewer: ```suggestion
   The :n:`No Variables` clause disables implicit generalization entirely.  This is the
```

seems more consistent

   default behavior.

> 

Reviewer: The new description should say that the `global` attribute is supported.

Author: Yeah, take a look at what I did.

   terms apparently identical but internally different in the Calculus of Inductive
   Constructions.

.. cmd:: Print {? Sorted } Universes {? Subgraph ( {* @qualid } ) } {? @string }
   :name: Print Universes

   This command can be used to print the constraints on the internal level
   of the occurrences of :math:`\Type` (see :ref:`Sorts`).

   The :n:`Subgraph` clause limits the printed graph to the requested names (adjusting
   constraints to preserve the implied transitive constraints between
   kept universes).

>    :n:`Sorted` makes each universe

Reviewer: ```suggestion
   The :n:`Sorted` clause makes each universe
```
